File: chudnovsky/src/test/java/com/apps/elliotgrin/chudnovsky/ChudnovskyAlgorithmThreadedTest.java
package com.apps.elliotgrin.chudnovsky;

import org.apfloat.Apfloat;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class ChudnovskyAlgorithmThreadedTest {

    private ChudnovskyAlgorithm alg = new ChudnovskyAlgorithm();
    
    @Test
    public void testSingleThreaded() {
        long precision = 200l;
        System.out.println("single-threaded Pi with precision of " + precision + ": " + alg.calculatePi(precision, true) + "\n");
    }

    @Test
    public void testMultiThreaded() {
        long precision = 200l;
        int numberOfThreads = 4;
        System.out.println("multi-threaded (" + numberOfThreads + " threads) Pi with precision of " + precision + ": " + alg.calculatePi(precision, numberOfThreads, true) + "\n");
    }

    @Test
    public void testCompareSingleToMultiThreaded() {
        for (long precision = 1; precision <= Math.pow(2, 15); precision *= 2l) {

            long startTime = System.nanoTime();
            Apfloat singleThreadedPi = alg.calculatePi(precision, true);
            long singleThreadedDuration = (System.nanoTime() - startTime);
            if (singleThreadedDuration > 0) {
                singleThreadedDuration /= 1000000;
            }
            System.out.println("single-threaded Pi with precision of " + precision + ": " + singleThreadedPi);
            System.out.println("execution time: " + singleThreadedDuration + " ms\n");

            for (int numberOfThreads = 1; numberOfThreads <= 8; numberOfThreads++) {

                startTime = System.nanoTime();
                Apfloat multiThreadedPi = alg.calculatePi(precision, numberOfThreads, true);
                long multiThreadedDuration = (System.nanoTime() - startTime);
                if (multiThreadedDuration > 0) {
                    multiThreadedDuration /= 1000000;
                }
                System.out.println("multi-threaded (" + numberOfThreads + " threads) Pi with precision of " + precision + ": " + multiThreadedPi);
                System.out.println("execution time: " + multiThreadedDuration + " ms");
                if(singleThreadedDuration > 0 && multiThreadedDuration > 0) {
                    System.out.println("% speed increase with " + numberOfThreads + " threads, from single threaded: " + String.format("%.2f", 100.0*(1.0 - ((double)multiThreadedDuration/singleThreadedDuration))) + "%\n");
                }

                assertEquals(singleThreadedPi, multiThreadedPi);
            }
        }
    }

}

File: chudnovsky/src/test/java/com/apps/elliotgrin/chudnovsky/ChudnovskyAlgorithmThreadedTest.kt
package com.apps.elliotgrin.chudnovsky

import org.junit.Test

import org.junit.Assert.assertEquals

class ChudnovskyAlgorithmThreadedTest {

    private val alg = ChudnovskyAlgorithm()

    @Test
    fun testSingleThreaded() {
        val precision = 200L
        println("single-threaded Pi with precision of " + precision + ": " + alg.calculatePi(precision, false) + "\n")
    }

    @Test
    fun testMultiThreaded() {
        val precision = 200L
        val numberOfThreads = 4
        println(
            "multi-threaded (" + numberOfThreads + " threads) Pi with precision of " + precision + ": " + alg.calculatePi(
                precision,
                false,
                numberOfThreads
            ) + "\n"
        )
    }

    @Test
    fun testCompareSingleToMultiThreaded() {
        var precision: Long = 1
        while (precision <= Math.pow(2.0, 15.0)) {

            var startTime = System.nanoTime()
            val singleThreadedPi = alg.calculatePi(precision, false)
            var singleThreadedDuration = System.nanoTime() - startTime
            if (singleThreadedDuration > 0) {
                singleThreadedDuration /= 1000000
            }
            println("single-threaded Pi with precision of $precision: $singleThreadedPi")
            println("execution time: $singleThreadedDuration ms\n")

            for (numberOfThreads in 1..8) {

                startTime = System.nanoTime()
                val multiThreadedPi = alg.calculatePi(precision, false, numberOfThreads)
                var multiThreadedDuration = System.nanoTime() - startTime
                if (multiThreadedDuration > 0) {
                    multiThreadedDuration /= 1000000
                }
                println("multi-threaded ($numberOfThreads threads) Pi with precision of $precision: $multiThreadedPi")
                println("execution time: $multiThreadedDuration ms")
                if (singleThreadedDuration > 0 && multiThreadedDuration > 0) {
                    println(
                        "% speed increase with $numberOfThreads threads, from single threaded: " + String.format(
                            "%.2f",
                            100.0 * (1.0 - multiThreadedDuration.toDouble() / singleThreadedDuration)
                        ) + "%\n"
                    )
                }

                assertEquals(singleThreadedPi, multiThreadedPi)
            }
            precision *= 2L
        }
    }

}

